### PR TITLE
api analytics and key events

### PR DESCRIPTION
**Pull request by andrewxhill**

### CODE REVIEW

File: api/billingd/service/service.go [hunk 1]
```diff
@@ -948,12 +936,23 @@ func (s *Service) handleUsage(ctx context.Context, cus *Customer, product Produc
...
 	if total > product.FreeQuotaSize && !cus.Billable {
 		now := time.Now().Unix()
+
+		summary := map[string]interface{}{
+			product.Key + "_name":       product.Name,
+			product.Key + "_usage":      total,
+			product.Key + "_units":      product.Units,
+			product.Key + "_free_quota": product.FreeQuotaSize,
+		}
+		go s.analytics.Update(cus.Key, cus.Email, false, summary)
```
Comment (andrewxhill) [inline api/billingd/service/service.go:946]: updates it on the user profile

File: api/billingd/service/service.go [hunk 2]
```diff
@@ -948,12 +936,23 @@ func (s *Service) handleUsage(ctx context.Context, cus *Customer, product Produc
...
+
 		if cus.GracePeriodStart == 0 {
 			cus.GracePeriodStart = now
 			update["grace_period_start"] = cus.GracePeriodStart
+			go s.analytics.NewEvent(cus.Key, "grace_period_start", false, summary)
 		}
 		deadline := cus.GracePeriodStart + int64(s.config.FreeQuotaGracePeriod.Seconds())
 		if now >= deadline {
+			go s.analytics.NewEvent(cus.Key, "grace_period_end", false, summary)
```
Comment (andrewxhill) [inline api/billingd/service/service.go:955]: an event to fire triggers such as email about status

File: cmd/hubd/docker-compose-dev.yml [hunk 3]
```diff
@@ -19,6 +19,9 @@ services:
       - HUB_ADDR_POWERGATE_API
       - HUB_GATEWAY_SUBDOMAINS
       - HUB_EMAIL_SESSION_SECRET=hubsession
+      - HUB_EMAIL_API_KEY
+      - HUB_SEGMENT_API_KEY
+      - HUB_SEGMENT_PREFIX
```
Comment (andrewxhill) [inline cmd/hubd/docker-compose-dev.yml:24]: helpful if testing local

File: cmd/hubd/main.go [hunk 4]
```diff
@@ -97,13 +97,13 @@ var (
...
+			"emailInviteTmpl": {
+				Key:      "email.invite_template",
+				DefValue: "2",
 			},
-			"emailDomain": {
-				Key:      "email.domain",
-				DefValue: "email.textile.io",
+			"emailConfirmTmpl": {
+				Key:      "email.confirm_template",
+				DefValue: "3",
```
Comment (andrewxhill) [inline cmd/hubd/main.go:106]: email template values, they can be updated on the service (2 and 3 are currently usable ones), but felt like config flags not hard-coded values in the code

**Comment by sanderpick** [comment]
This is awesome! Will take a look in a bit.

**Review by sanderpick — APPROVED** [review]
Looks good! 

- I still think we should rename the billing service to "usage", then we could expose the analytics APIs there. That would keep the segment stuff consolidated there. Especially when Hub becomes just a service runner + interceptors. Could do in a follow up PR if you agree.

File: analytics/analytics.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+	api    segment.Client
+	prefix string
+	debug  bool
+}
+
+// NewClient return a segment client.
+func NewClient(segmentAPIKey, prefix string, debug bool) (*Client, error) {
+	var api segment.Client
+	var err error
+	if segmentAPIKey != "" {
```
[the remark this answers is not in the document]
Comment (andrewxhill) [reply analytics/analytics.go:26]: i was just letting it gracefully continue. if you don't need to run it, just don't give a key. this is nice so all those future calls to the methods don't have to have checks if it had error previously, they just make the call as usual.

**Comment by andrewxhill** [comment]
okay, i think that's finally gtg. moving it all to 'usage' i think makes sense, sure!

File: api/billingd/service/service.go [hunk 6]
```diff
@@ -991,6 +990,14 @@ func (s *Service) reportUsage() error {
 		if err := cursor.Decode(&doc); err != nil {
 			return fmt.Errorf("decoding customer: %v", err)
 		}
+		go s.analytics.Update(doc.Key, doc.Email, false, map[string]interface{}{
```
[the remark this answers is not in the document]
Comment (sanderpick) [reply api/billingd/service/service.go:993]: Nice, sgtm

**Comment by sanderpick** [comment]
Nice, LGTM!

**Pull request merged**